elastic/kibana
# [Stack Monitoring] Filter out partial Logstash pipeline documents

## Summary

Fixes https://github.com/elastic/kibana/issues/133228

This is the Kibana side of a series of fixes that deal with a race condition whereby Metricbeat ingests partial Logstash data due to Logstash returning partial results from it's APIs when the pipeline setup process hasn't finished. This results in the Stack Monitoring pipeline UI erroring out as bad data is returned from the aggregations used to determine versions (based off hashes).

This little chunk of code:

```ts
return pipelineHash
    ? versions.find(({ hash }) => hash === pipelineHash) ?? versions[0]
    : versions[0];
```

would invariably return `versions[0]` (as the UI doesn't request a specific hash) which is the bad result with `hash: ''` (hash of an empty string, which is invalid).

## Testing

- Checkout `main`

- Setup and run Metricbeat with the Logstash module enabled. Set the `period` to `1s` in your config (this way we will definitely poll at the point where setup hasn't completed). This is a race condition so naturally dependant on timing, but at least in my testing where I used Docker for Metricbeat and Logstash this always worked. 

- Now, start Logstash.

- At this point one (at least) bad document should be ingested whereby hash is empty (`hash: ''`)

- Observe the pipeline UI is broken (see here for more details: https://github.com/elastic/kibana/issues/130217)

- Now checkout this branch, observe that the UI works as we filter out the broken data, and we use the correct `hash` from the data ingested after.
